# **Week 2 Applied Session: Introduction to DataWrangling with Pandas**

![](https://media.licdn.com/dms/image/D5612AQEdvrs4ha4KAQ/article-cover_image-shrink_600_2000/0/1693676526923?e=2147483647&v=beta&t=aegFlNZu0P_4UKcfh4ZTol_MmcIQzqoZx5tOKKMkI1E)

Pandas is a strange name, kind of an acronym: Python, Numerical, Data Analysis?

Because pandas is an external library you need to import it. There are several ways that you will see imports done:
- import pandas
- from pandas import tools
- import pandas as `pd`

The first is the same as `from pandas import *` where star means all (that's right, the same as SQL)

The second imports a part of pandas only, a sublibrary called *tools*

The third is a renaming, or alias, `pd` is common (you could call pandas `xyz` but you'd be on your own).

You could leave out the `import as` and just type `pandas` every time but it becomes more useful for longer names e.g. `import matplotlib.pyplot as plt`

So, for any code following (if the above imports work), `plt` would mean `matplotlib.pyplot`

This, by the way, is a Python Notebook, select cells (this one is text, below is code) then `SHIFT-ENTER` to run sequentially

The following scripts should work with both Python2 and Python 3!

In [159]:
# import libraries first
import pandas as pd
import numpy as np # Numberical Python

## **1. The Pandas DataFrame**

In [160]:
# so try pandas.DataFrame()
pd.DataFrame()

""


So.. no errors, seems to have worked, but what's in the DataFrame? (nothing)

**note**: Python is case sensitive: `DataFrame` is not the same as `Dataframe` or `dataframe`

In [161]:
pd.DataFrame([2,4,6,8])

,0
0,2
1,4
2,6
3,8


In [162]:
# aha, better but this is temporary, if you want to use the data you need to save it, so create a variable
df = pd.DataFrame([2,4,6,8])

In [163]:
# but now there's no output... can't win
# use the variable to see the data
df

,0
0,2
1,4
2,6
3,8


**Note**: the column titles are ` ` and `0`

And another note: Python is one of those `0` index languages, we have 4 items `(2,4,6,8)` but they are found at `0,1,2,3` viz:

In [164]:
# you can get the values with its index:
df[0][1] # column 0, item 1

np.int64(4)

In [165]:
# rename the column
df.columns.name = "Index"
df

Index,0
0,2
1,4
2,6
3,8


In [166]:
# You can also use pandas to create an series of datetime objects. Let's make one for the week beginning January 25th, 2015:
dates = pd.date_range('20150125', periods=7)

dates

DatetimeIndex(['2015-01-25', '2015-01-26', '2015-01-27', '2015-01-28',
               '2015-01-29', '2015-01-30', '2015-01-31'],
              dtype='datetime64[us]', freq='D')

Now we'll create a DataFrame using the dates array as our index, fill it with some random values using numpy, and give the columns some labels.

Note that `randn(7,5)` below matches the 7 dates (rows) and 5 names (columns)

(Otherwise it wouldn't work, try changing 5 to 6...)

In [167]:
df = pd.DataFrame(np.random.randn(7,5), index=dates, columns=['Adam','Bob','Carla','Dave','Eve'])
df

,Adam,Bob,Carla,Dave,Eve
2015-01-25,1.068028,-1.555162,-1.119964,-0.443243,-0.786383
2015-01-26,1.461746,-0.331045,-0.925439,0.449471,-0.336225
2015-01-27,0.073635,-0.278493,0.674920,-0.478516,0.573509
2015-01-28,0.351914,0.379390,0.594328,0.399870,0.567707
2015-01-29,-1.927059,1.030311,0.575659,-1.244948,1.996933
2015-01-30,-0.459636,2.195240,0.196718,1.136544,1.298310
2015-01-31,0.677911,-0.606260,0.572903,-0.742959,0.991342


DataFrames are more flexible than that, both in terms of what you can store in them and what you can do with them.

It can also be useful to know how to create a DataFrame from a dict of objects.

This comes in particularly handy when working with JSON-like structures.

In [168]:
df2 = pd.DataFrame({ 'A' : np.random.random_sample(4), # 4 random numbers
                     'B' : pd.Timestamp('20130102'), # 4 dates, note pandas autofills
                     'C' : pd.date_range('20150125',periods = 4), # 4 dates in a range
                     'D' : ['a','b','c','d'], # letters
                     'E' : ["cat","dog","mouse","parrot"], # text/string
                     'F' : 'copy'}) # note pandas autofills

df2

,A,B,C,D,E,F
0,0.407571,2013-01-02,2015-01-25,a,cat,copy
1,0.841604,2013-01-02,2015-01-26,b,dog,copy
2,0.881396,2013-01-02,2015-01-27,c,mouse,copy
3,0.000205,2013-01-02,2015-01-28,d,parrot,copy




---



## **2. Exploring the data in a DataFrame**

Let's import the UFO sightings dataset via URL. We can access the data types of each column in a DataFrame as follows:

In [169]:
ufo = pd.read_csv('http://bit.ly/uforeports')

We can display the index, columns and the underlyinig numpy data separately:

In [170]:
ufo.index

RangeIndex(start=0, stop=18241, step=1)

In [171]:
ufo.columns

Index(['City', 'Colors Reported', 'Shape Reported', 'State', 'Time'], dtype='str')

We can accesss the data types of each column:

In [172]:
ufo.dtypes

City               str
Colors Reported    str
Shape Reported     str
State              str
Time               str
dtype: object

In [173]:
ufo.values

array([['Ithaca', nan, 'TRIANGLE', 'NY', '6/1/1930 22:00'],
       ['Willingboro', nan, 'OTHER', 'NJ', '6/30/1930 20:00'],
       ['Holyoke', nan, 'OVAL', 'CO', '2/15/1931 14:00'],
       ...,
       ['Eagle River', nan, nan, 'WI', '12/31/2000 23:45'],
       ['Eagle River', 'RED', 'LIGHT', 'WI', '12/31/2000 23:45'],
       ['Ybor', nan, 'OVAL', 'FL', '12/31/2000 23:59']],
      shape=(18241, 5), dtype=object)

We can get the size of the data using `shape`:

In [174]:
ufo.shape

(18241, 5)

We can get a quick statistical summary of the data using `describe()` function:

In [175]:
ufo.describe()

,City,Colors Reported,Shape Reported,State,Time
count,18215,2882,15597,18241,18241
unique,6475,27,27,52,16145
top,Seattle,RED,LIGHT,CA,11/16/1999 19:00
freq,187,780,2803,2529,27


We can have a look at the data in first five rows using `head()` function:

In [176]:
ufo.head()

,City,Colors Reported,Shape Reported,State,Time
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00
3,Abilene,NaN,DISK,KS,6/1/1931 13:00
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00


We can indicate how many row to return by specifying an integer:

In [177]:
ufo.head(20)

,City,Colors Reported,Shape Reported,State,Time
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00
3,Abilene,NaN,DISK,KS,6/1/1931 13:00
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00
5,Valley City,NaN,DISK,ND,9/15/1934 15:30
6,Crater Lake,NaN,CIRCLE,CA,6/15/1935 0:00
7,Alma,NaN,DISK,MI,7/15/1936 0:00
8,Eklutna,NaN,CIGAR,AK,10/15/1936 17:00
9,Hubbard,NaN,CYLINDER,OR,6/15/1937 0:00


We can also have the last five rows using `tail()` function, check the index numbers:

In [178]:
ufo.tail()

,City,Colors Reported,Shape Reported,State,Time
18236,Grant Park,NaN,TRIANGLE,IL,12/31/2000 23:00
18237,Spirit Lake,NaN,DISK,IA,12/31/2000 23:00
18238,Eagle River,NaN,NaN,WI,12/31/2000 23:45
18239,Eagle River,RED,LIGHT,WI,12/31/2000 23:45
18240,Ybor,NaN,OVAL,FL,12/31/2000 23:59


We can focus on a specific column:

In [179]:
ufo['City']

0                      Ithaca
1                 Willingboro
2                     Holyoke
3                     Abilene
4        New York Worlds Fair
                 ...         
18236              Grant Park
18237             Spirit Lake
18238             Eagle River
18239             Eagle River
18240                    Ybor
Name: City, Length: 18241, dtype: str

We can select a subset of rows by integer indexing:

In [180]:
ufo[1:3]

,City,Colors Reported,Shape Reported,State,Time
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00


**Note**：Only the rows wiht index 1 and 2 returned.

We can also select rows by specific values:

In [181]:
ufo[ufo.State == 'NJ'][['City','State']]

,City,State
1,Willingboro,NJ
141,Newark,NJ
147,Sandy Hook,NJ
255,Keansburg,NJ
311,Red Bank,NJ
...,...,...
17961,Freehold,NJ
17976,East Brunswick,NJ
18006,West Allenhurst,NJ
18023,Elizabeth,NJ


**Note**: `ufo.State == 'NJ'` is an example of using conditional indexing. We can also have other conditions like `<`, `>`, `<=`, `>=` or `!=` (not equal).

In [182]:
ufo[['City','State']][ufo.State != 'NJ']

,City,State
0,Ithaca,NY
2,Holyoke,CO
3,Abilene,KS
4,New York Worlds Fair,NY
5,Valley City,ND
...,...,...
18236,Grant Park,IL
18237,Spirit Lake,IA
18238,Eagle River,WI
18239,Eagle River,WI


What is the difference between the following two functions `loc` and `iloc`?

In [183]:
ufo.loc[:,'City':'State'].head()

,City,Colors Reported,Shape Reported,State
0,Ithaca,NaN,TRIANGLE,NY
1,Willingboro,NaN,OTHER,NJ
2,Holyoke,NaN,OVAL,CO
3,Abilene,NaN,DISK,KS
4,New York Worlds Fair,NaN,LIGHT,NY


In [184]:
ufo.iloc[1:6,2:6]

,Shape Reported,State,Time
1,OTHER,NJ,6/30/1930 20:00
2,OVAL,CO,2/15/1931 14:00
3,DISK,KS,6/1/1931 13:00
4,LIGHT,NY,4/18/1933 19:00
5,DISK,ND,9/15/1934 15:30


Enter your answer here....



---



## **Task 1: Load data and get the basic information**

In this task, you are asked to load a data file from Google Drive with your Monash account. Open the file to create a Pandas dataframe and explore it using the functions introduced above and see what information you can get from the data.

### **Connect with your Google Drive to access files**

In [185]:
# Note: google.colab is not available in local Jupyter environments
# This cell is for Google Colab only
# from google.colab import drive
# drive.mount('/content/drive')
print("Running in local environment - no need to mount Google Drive")

Running in local environment - no need to mount Google Drive


After you run the cell, then it asks you to click on a url and login in order to give premissions to Colab. If you successfully followed the steps, you should now see a drive folder in the left pane of this notebook. see below figure.

![](https://drive.google.com/uc?export=download&id=1B2sooICEr_QDLEyFHOSwIQO89LODLfAq)

If you click on it, you should be able to see the "FIT5196_S1_2025" shared drive. If you are unable to see that, let us know ASAP. But if you can see it, then it means that now this notebook have access to everything on that shared drive. Let's read the `xmart` data from there.

In [186]:
xmart = pd.read_csv('xmart.csv',skiprows=1)


Now, it is your turn to write Python codes and try to find out:
1. How many records in the dataset?
2. How many attributes in the dataset? What are they?
3. What is the data type for each of the attribute?
4. Without any description provided, can you summarize what information contains in this dataset?

**Note**: Don't forget to use markdown to explain your findings.

In [187]:
# How many records in the dataset?
xmart.shape[0]

940

In [188]:
# How many attributes in the dataset? What are they?
xmart.shape[1], xmart.columns.tolist()

(4, ['Country', 'Data Source', 'Beverage Types', ' 2012'])

In [189]:
# What is the data type for each of the attribute?
xmart.dtypes

Country               str
Data Source           str
Beverage Types        str
 2012             float64
dtype: object

In [190]:
# Without any description provided, can you summarize what information contains in this dataset?
xmart.info()

<class 'pandas.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country         940 non-null    str    
 1   Data Source     940 non-null    str    
 2   Beverage Types  940 non-null    str    
 3    2012           940 non-null    float64
dtypes: float64(1), str(3)
memory usage: 29.5 KB




---



## **3. Editing data in DataFrame**

We can apply basic editing operations to DataFrame objects, such as updating, deleting, duplication, adding new columns, insert new rows, and etc.

### **3.1 Dealing with missing values (simplest ways)**

Continue with ufo data, we are looking for the data rows having missing city data and simply remove them.

In [191]:
ufo[10:20]

,City,Colors Reported,Shape Reported,State,Time
10,Fontana,NaN,LIGHT,CA,8/15/1937 21:00
11,Waterloo,NaN,FIREBALL,AL,6/1/1939 20:00
12,Belton,RED,SPHERE,SC,6/30/1939 20:00
13,Keokuk,NaN,OVAL,IA,7/7/1939 2:00
14,Ludington,NaN,DISK,MI,6/1/1941 13:00
15,Forest Home,NaN,CIRCLE,CA,7/2/1941 11:30
16,Los Angeles,NaN,NaN,CA,2/25/1942 0:00
17,Hapeville,NaN,NaN,GA,6/1/1942 22:30
18,Oneida,NaN,RECTANGLE,TN,7/15/1942 1:00
19,Bering Sea,RED,OTHER,AK,4/30/1943 23:00


In [192]:
# check whether the values is NaN or not
ufo.isna()

,City,Colors Reported,Shape Reported,State,Time
0,False,True,False,False,False
1,False,True,False,False,False
2,False,True,False,False,False
3,False,True,False,False,False
4,False,True,False,False,False
...,...,...,...,...,...
18236,False,True,False,False,False
18237,False,True,False,False,False
18238,False,True,True,False,False
18239,False,False,False,False,False


In [193]:
# find out how many missing values in each attribute
ufo.isna().sum()

City                  26
Colors Reported    15359
Shape Reported      2644
State                  0
Time                   0
dtype: int64

In [194]:
# list all the rows with missing values
ufo[ufo.isna().any(axis=1)]

,City,Colors Reported,Shape Reported,State,Time
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00
3,Abilene,NaN,DISK,KS,6/1/1931 13:00
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00
...,...,...,...,...,...
18235,Fountain Hills,NaN,NaN,AZ,12/31/2000 23:00
18236,Grant Park,NaN,TRIANGLE,IL,12/31/2000 23:00
18237,Spirit Lake,NaN,DISK,IA,12/31/2000 23:00
18238,Eagle River,NaN,NaN,WI,12/31/2000 23:45


**Note**: The ufo dataset has 18,241 rows in total, but 15,755  of them have at least one missing value. We can simply remove all of them to keep the rows with complete data for next step analysis. To keep the original data, we create a new data frame to store the subset of data rows with complete values.

In [195]:
# remove data rows with missing values
ufo1 = ufo.dropna()
ufo1.head(20)

,City,Colors Reported,Shape Reported,State,Time
12,Belton,RED,SPHERE,SC,6/30/1939 20:00
19,Bering Sea,RED,OTHER,AK,4/30/1943 23:00
36,Portsmouth,RED,FORMATION,VA,7/10/1945 1:30
44,Blairsden,GREEN,SPHERE,CA,6/30/1946 19:00
82,San Jose,BLUE,CHEVRON,CA,7/15/1947 21:00
84,Modesto,BLUE,DISK,CA,8/8/1947 22:00
91,Scipio,RED,SPHERE,IN,5/10/1948 19:00
111,Tarrant City,ORANGE,CIRCLE,AL,8/15/1949 22:00
129,Napa,GREEN,DISK,CA,6/10/1950 0:00
138,Coeur d'Alene,ORANGE,CIGAR,ID,7/2/1950 13:00


In [196]:
ufo1.isna().sum()

City               0
Colors Reported    0
Shape Reported     0
State              0
Time               0
dtype: int64

In this way, only a very small part of the data is kept and we may not have enough data to work out any useful knowledge. Instead of removing all the missing values, we can replace missing parts with some specific values. For example, set them all to zero.

In [197]:
ufo2 = ufo.fillna(value=0)
ufo2.head(20)

,City,Colors Reported,Shape Reported,State,Time
0,Ithaca,0,TRIANGLE,NY,6/1/1930 22:00
1,Willingboro,0,OTHER,NJ,6/30/1930 20:00
2,Holyoke,0,OVAL,CO,2/15/1931 14:00
3,Abilene,0,DISK,KS,6/1/1931 13:00
4,New York Worlds Fair,0,LIGHT,NY,4/18/1933 19:00
5,Valley City,0,DISK,ND,9/15/1934 15:30
6,Crater Lake,0,CIRCLE,CA,6/15/1935 0:00
7,Alma,0,DISK,MI,7/15/1936 0:00
8,Eklutna,0,CIGAR,AK,10/15/1936 17:00
9,Hubbard,0,CYLINDER,OR,6/15/1937 0:00


Oops... we have nominal attributes, not numeric. Zeros do not work!!!

First, let's copy the DataFrame to have a new one to work on.

In [198]:
ufo3 = ufo.copy()
ufo3[10:20]

,City,Colors Reported,Shape Reported,State,Time
10,Fontana,NaN,LIGHT,CA,8/15/1937 21:00
11,Waterloo,NaN,FIREBALL,AL,6/1/1939 20:00
12,Belton,RED,SPHERE,SC,6/30/1939 20:00
13,Keokuk,NaN,OVAL,IA,7/7/1939 2:00
14,Ludington,NaN,DISK,MI,6/1/1941 13:00
15,Forest Home,NaN,CIRCLE,CA,7/2/1941 11:30
16,Los Angeles,NaN,NaN,CA,2/25/1942 0:00
17,Hapeville,NaN,NaN,GA,6/1/1942 22:30
18,Oneida,NaN,RECTANGLE,TN,7/15/1942 1:00
19,Bering Sea,RED,OTHER,AK,4/30/1943 23:00


In [199]:
ufo3['Colors Reported'].fillna('BLUE')
ufo3[10:20]

,City,Colors Reported,Shape Reported,State,Time
10,Fontana,NaN,LIGHT,CA,8/15/1937 21:00
11,Waterloo,NaN,FIREBALL,AL,6/1/1939 20:00
12,Belton,RED,SPHERE,SC,6/30/1939 20:00
13,Keokuk,NaN,OVAL,IA,7/7/1939 2:00
14,Ludington,NaN,DISK,MI,6/1/1941 13:00
15,Forest Home,NaN,CIRCLE,CA,7/2/1941 11:30
16,Los Angeles,NaN,NaN,CA,2/25/1942 0:00
17,Hapeville,NaN,NaN,GA,6/1/1942 22:30
18,Oneida,NaN,RECTANGLE,TN,7/15/1942 1:00
19,Bering Sea,RED,OTHER,AK,4/30/1943 23:00


Why do we stil have the missing values? Nothing happened?

Let's try it again:

In [200]:
# Change the null values with 'BLUE' in 'Colors Reported'
ufo3['Colors Reported'].fillna('BLUE', inplace=True)
ufo3[10:20]

/var/folders/6s/bb6mgp0x50q3fjlqfsfpffjm0000gn/T/ipykernel_68489/3793977870.py:2: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  ufo3['Colors Reported'].fillna('BLUE', inplace=True)


,City,Colors Reported,Shape Reported,State,Time
10,Fontana,NaN,LIGHT,CA,8/15/1937 21:00
11,Waterloo,NaN,FIREBALL,AL,6/1/1939 20:00
12,Belton,RED,SPHERE,SC,6/30/1939 20:00
13,Keokuk,NaN,OVAL,IA,7/7/1939 2:00
14,Ludington,NaN,DISK,MI,6/1/1941 13:00
15,Forest Home,NaN,CIRCLE,CA,7/2/1941 11:30
16,Los Angeles,NaN,NaN,CA,2/25/1942 0:00
17,Hapeville,NaN,NaN,GA,6/1/1942 22:30
18,Oneida,NaN,RECTANGLE,TN,7/15/1942 1:00
19,Bering Sea,RED,OTHER,AK,4/30/1943 23:00


### **3.2 Create a new column**

Let's create a new column with the combined City and State place names, called `place` with an empty string in every row. This isn't absolutely necessary when using proper Pandas methods but for the demonstration it will make it more straight forward.

In [201]:
ufo4 = ufo.copy()
ufo4['place']=''
ufo4

,City,Colors Reported,Shape Reported,State,Time,place
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00,
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00,
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00,
3,Abilene,NaN,DISK,KS,6/1/1931 13:00,
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00,
...,...,...,...,...,...,...
18236,Grant Park,NaN,TRIANGLE,IL,12/31/2000 23:00,
18237,Spirit Lake,NaN,DISK,IA,12/31/2000 23:00,
18238,Eagle River,NaN,NaN,WI,12/31/2000 23:45,
18239,Eagle River,RED,LIGHT,WI,12/31/2000 23:45,


**Note**: By default, the new column is always added at the end

Before we combine the city and state, we need to check whether there are missing values in these two columns. From above, we know the `City` column has 26 rows with missing values and the `State` is complete. So we need to fill the `City` column before merging.

In [202]:
ufo4['City'].fillna('No city', inplace=True)

/var/folders/6s/bb6mgp0x50q3fjlqfsfpffjm0000gn/T/ipykernel_68489/721252561.py:1: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  ufo4['City'].fillna('No city', inplace=True)


0                      Ithaca
1                 Willingboro
2                     Holyoke
3                     Abilene
4        New York Worlds Fair
                 ...         
18236              Grant Park
18237             Spirit Lake
18238             Eagle River
18239             Eagle River
18240                    Ybor
Name: City, Length: 18241, dtype: str

In [203]:
ufo4['place'] = ufo4['City'] + ', ' + ufo4['State']
ufo4

,City,Colors Reported,Shape Reported,State,Time,place
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00,"Ithaca, NY"
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00,"Willingboro, NJ"
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00,"Holyoke, CO"
3,Abilene,NaN,DISK,KS,6/1/1931 13:00,"Abilene, KS"
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00,"New York Worlds Fair, NY"
...,...,...,...,...,...,...
18236,Grant Park,NaN,TRIANGLE,IL,12/31/2000 23:00,"Grant Park, IL"
18237,Spirit Lake,NaN,DISK,IA,12/31/2000 23:00,"Spirit Lake, IA"
18238,Eagle River,NaN,NaN,WI,12/31/2000 23:45,"Eagle River, WI"
18239,Eagle River,RED,LIGHT,WI,12/31/2000 23:45,"Eagle River, WI"


We can use a `for` loop to achieve the same result.

Before we apply any operations that take use of index, remember to check the valid index range to avoid any errors.

In [204]:
ufo4.index

RangeIndex(start=0, stop=18241, step=1)

In [205]:
ufo4['address']=''
ufo4

,City,Colors Reported,Shape Reported,State,Time,place,address
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00,"Ithaca, NY",
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00,"Willingboro, NJ",
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00,"Holyoke, CO",
3,Abilene,NaN,DISK,KS,6/1/1931 13:00,"Abilene, KS",
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00,"New York Worlds Fair, NY",
...,...,...,...,...,...,...,...
18236,Grant Park,NaN,TRIANGLE,IL,12/31/2000 23:00,"Grant Park, IL",
18237,Spirit Lake,NaN,DISK,IA,12/31/2000 23:00,"Spirit Lake, IA",
18238,Eagle River,NaN,NaN,WI,12/31/2000 23:45,"Eagle River, WI",
18239,Eagle River,RED,LIGHT,WI,12/31/2000 23:45,"Eagle River, WI",


In [206]:
for i in ufo4.index:
    # check whether the value in 'City' is null or not
    if pd.isna(ufo4.iloc[i,0]):
        print('Missing value in City')

Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City
Missing value in City


In [207]:
# Using a for loop to create each entry in turn

# for i in ufo4.index:
#     ufo4.iloc[i,6] = ufo4.iloc[i,0] + ', ' + ufo4.iloc[i,3]

for i in ufo4.index:
    ufo4.iloc[i,6] = str(ufo4.iloc[i,0]) + ', ' + str(ufo4.iloc[i,3])

In [208]:
ufo4


,City,Colors Reported,Shape Reported,State,Time,place,address
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00,"Ithaca, NY","Ithaca, NY"
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00,"Willingboro, NJ","Willingboro, NJ"
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00,"Holyoke, CO","Holyoke, CO"
3,Abilene,NaN,DISK,KS,6/1/1931 13:00,"Abilene, KS","Abilene, KS"
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00,"New York Worlds Fair, NY","New York Worlds Fair, NY"
...,...,...,...,...,...,...,...
18236,Grant Park,NaN,TRIANGLE,IL,12/31/2000 23:00,"Grant Park, IL","Grant Park, IL"
18237,Spirit Lake,NaN,DISK,IA,12/31/2000 23:00,"Spirit Lake, IA","Spirit Lake, IA"
18238,Eagle River,NaN,NaN,WI,12/31/2000 23:45,"Eagle River, WI","Eagle River, WI"
18239,Eagle River,RED,LIGHT,WI,12/31/2000 23:45,"Eagle River, WI","Eagle River, WI"


We have the same values in both `place` and `address`. But, which way is better?

### **3.3 Timing it**

The notebook's magic `%%timeit` will run the cell 1000 times and get the 3 quickest times. We can use it to record the time and then do the comparison.

In [209]:
ufo5 = ufo.copy()
ufo5['City'].fillna('No city', inplace=True)
ufo5

/var/folders/6s/bb6mgp0x50q3fjlqfsfpffjm0000gn/T/ipykernel_68489/3128832350.py:2: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  ufo5['City'].fillna('No city', inplace=True)


,City,Colors Reported,Shape Reported,State,Time
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00
3,Abilene,NaN,DISK,KS,6/1/1931 13:00
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00
...,...,...,...,...,...
18236,Grant Park,NaN,TRIANGLE,IL,12/31/2000 23:00
18237,Spirit Lake,NaN,DISK,IA,12/31/2000 23:00
18238,Eagle River,NaN,NaN,WI,12/31/2000 23:45
18239,Eagle River,RED,LIGHT,WI,12/31/2000 23:45


In [210]:
ufo5['place']=''
ufo5['address']=''
ufo5.head()

,City,Colors Reported,Shape Reported,State,Time,place,address
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00,,
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00,,
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00,,
3,Abilene,NaN,DISK,KS,6/1/1931 13:00,,
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00,,


In [211]:
%%timeit

ufo5['place'] = ufo5['City'] + ', ' + ufo5['State']

1.82 ms ± 32.5 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [212]:
%%timeit

for i in ufo5.index:
    ufo5.iloc[i,6] = str(ufo5.iloc[i,0]) + ', ' + str(ufo5.iloc[i,3])

834 ms ± 3.84 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


You can run the above codes many times, they may give different time, but using `for` loop is a much slower method.

**Note**: Pandas is based on numpy arrays, so try everything you can to aviod iterating over rows.

### **3.4 Delete columns**

We only need to keep one column for merged place, we can easily drop one.

In [213]:
ufo5.drop(columns=['address'], inplace=True)
ufo5.head()

,City,Colors Reported,Shape Reported,State,Time,place
0,Ithaca,NaN,TRIANGLE,NY,6/1/1930 22:00,"Ithaca, NY"
1,Willingboro,NaN,OTHER,NJ,6/30/1930 20:00,"Willingboro, NJ"
2,Holyoke,NaN,OVAL,CO,2/15/1931 14:00,"Holyoke, CO"
3,Abilene,NaN,DISK,KS,6/1/1931 13:00,"Abilene, KS"
4,New York Worlds Fair,NaN,LIGHT,NY,4/18/1933 19:00,"New York Worlds Fair, NY"




---



## **Task 2: Reproducing the data wrangling process**

Load an `Air Crashes` data, and try to answer the following questions:
1. Give a summary of the data, including size, attributes, data types
2. Does the dataset contain missing value? What are you going to deal with them?
3. Check if there are any columns can be merged together? Apply the merging operation.
4. Remove the column(s) that contains duplicated information after merging.
5. Find out a subset that records air crashes with survivors.

In [214]:
aircrash = pd.read_csv('AirCrashes.csv')
aircrash.head()

,Incident American Airlines Flight 11 involving a Boeing 767-223ER in 2001
Casualties,Extremely High
Total Dead,1692
Crew,11
Passengers,81
Ground,1600


In [215]:
# Read raw file
with open("AirCrashes.csv", "r", encoding="utf-8", errors="ignore") as f:
    raw = f.read().split("\n")

# print first 10 lines
for line in raw[:20]:
    print(line)


Incident American Airlines Flight 11 involving a Boeing 767-223ER in 2001
Casualties,Extremely High
Total Dead,1692
Crew,11
Passengers,81
Ground,1600
Notes,No survivors
Type,INH
Reason,Attack
Location,New York - New York - US
Country,US
Phase,ENR
Date,2001-09-11
Latitude,40.7143528
Longitude,-74.0059731
Circumstances,Good Visibility by Day

Incident United Airlines Flight 175 involving a Boeing 767-222 in 2001
Casualties,Extremely High


In [216]:
# Split blocks by empty line
blocks = []
block = []
for line in raw:
    if line.strip() == "":
        if block:
            blocks.append(block)
            block = []
    else:
        block.append(line.strip())
if block:
    blocks.append(block)

blocks[:5]  # Display the first 5 blocks for inspection

[['Incident American Airlines Flight 11 involving a Boeing 767-223ER in 2001',
  'Casualties,Extremely High',
  'Total Dead,1692',
  'Crew,11',
  'Passengers,81',
  'Ground,1600',
  'Notes,No survivors',
  'Type,INH',
  'Reason,Attack',
  'Location,New York - New York - US',
  'Country,US',
  'Phase,ENR',
  'Date,2001-09-11',
  'Latitude,40.7143528',
  'Longitude,-74.0059731',
  'Circumstances,Good Visibility by Day'],
 ['Incident United Airlines Flight 175 involving a Boeing 767-222 in 2001',
  'Casualties,Extremely High',
  'Total Dead,965',
  'Crew,9',
  'Passengers,56',
  'Ground,900',
  'Notes,No survivors',
  'Type,INH',
  'Reason,Attack',
  'Location,New York - New York - US',
  'Country,USA',
  'Phase,ENR',
  'Date,2001-09-11',
  'Latitude,40.7143528',
  'Longitude,-74.0059731',
  'Circumstances,Good Visibility by Day'],
 ['Incident Pan Am Flight 1736 and KLM Flight 4805 involving a Boeing 747-121 and\xa0Boeing 747-206B in 1977',
  'Casualties,Extremely High',
  'Total Dead,583

In [217]:
# Convert each block into a dictionary
records = []
for b in blocks:
    rec = {}
    for line in b:
        if "," in line:
            key, val = line.split(",", 1)
            rec[key.strip()] = val.strip()
        else:
            # First line (incident description) without comma
            rec["Incident"] = line
    records.append(rec)

# print the first 5 records
for record in records[:5]:
    print(record)

{'Incident': 'Incident American Airlines Flight 11 involving a Boeing 767-223ER in 2001', 'Casualties': 'Extremely High', 'Total Dead': '1692', 'Crew': '11', 'Passengers': '81', 'Ground': '1600', 'Notes': 'No survivors', 'Type': 'INH', 'Reason': 'Attack', 'Location': 'New York - New York - US', 'Country': 'US', 'Phase': 'ENR', 'Date': '2001-09-11', 'Latitude': '40.7143528', 'Longitude': '-74.0059731', 'Circumstances': 'Good Visibility by Day'}
{'Incident': 'Incident United Airlines Flight 175 involving a Boeing 767-222 in 2001', 'Casualties': 'Extremely High', 'Total Dead': '965', 'Crew': '9', 'Passengers': '56', 'Ground': '900', 'Notes': 'No survivors', 'Type': 'INH', 'Reason': 'Attack', 'Location': 'New York - New York - US', 'Country': 'USA', 'Phase': 'ENR', 'Date': '2001-09-11', 'Latitude': '40.7143528', 'Longitude': '-74.0059731', 'Circumstances': 'Good Visibility by Day'}
{'Incident': 'Incident Pan Am Flight 1736 and KLM Flight 4805 involving a Boeing 747-121 and\xa0Boeing 747-20

In [218]:
# Create dataframe
df = pd.DataFrame(records)
df

,Incident,Casualties,Total Dead,Crew,Passengers,Ground,Notes,Type,Reason,Location,Country,Phase,Date,Latitude,Longitude,Circumstances
0,Incident American Airlines Flight 11 involving...,Extremely High,1692,11,81,1600,No survivors,INH,Attack,New York - New York - US,US,ENR,2001-09-11,40.7143528,-74.0059731,Good Visibility by Day
1,Incident United Airlines Flight 175 involving ...,Extremely High,965,9,56,900,No survivors,INH,Attack,New York - New York - US,USA,ENR,2001-09-11,40.7143528,-74.0059731,Good Visibility by Day
2,Incident Pan Am Flight 1736 and KLM Flight 480...,Extremely High,583,23,560,0,Some survivors,COM,Accident,Tenerife - Spain,Spain,TOF,1977-03-27,28.2915637,-16.6291304,Bad Visibility by Day
3,Incident Japan Airlines Flight 123 involving a...,Extremely High,520,15,505,0,Some survivors,COM,Accident,Ueno - Japan,Japan,ENR,1985-08-12,35.7089461,139.7742683,Bad Visibility by Night
4,Incident Saudi Arabian Flight 763 and involvin...,Extremely High,349,33,316,0,No survivors,,Accident,Charkhi Dadri - India,India,ENR,1996-11-12,28.6,76.2667,Bad Visibility by Night
5,Incident Turkish Airlines Flight 981 involving...,Extremely High,346,11,335,0,No survivors,COM,Accident,Fontaine-Chaalis - France,France,ENR,1974-03-03,49.171795,2.684157,Bad Visibility by Night
6,Incident Air India Flight 182 involving a Boei...,Extremely High,329,22,307,0,No survivors,INB,Attack,Atlantic Ocean - Cork - Ireland,Ireland,ENR,1985-06-23,51.8968917,-8.4863157,Bad Visibility by Night
7,Incident Saudia Flight 163 involving a Lockhee...,Extremely High,301,14,287,0,No survivors,COM,Accident,Riyadh - Saudi Arabia,Saudi Arabia,ENR,1980-08-19,24.7116667,46.7241667,Bad Visibility by Night
8,Incident Iran Air Flight 655 involving a Airbu...,Extremely High,290,16,274,0,No survivors,EXG,Attack,Strait of Hormuz - Iran,Iran,ENR,1988-07-03,26.5944754,56.4719928,Bad Visibility by Night
9,Incident Iranian Air Force (15-2280) involving...,Extremely High,275,18,257,0,No survivors,MIL,Accident,Kerman - Iran,Iran,ENR,2003-02-19,30.28027,57.06702,Bad Visibility by Night


In [219]:
# check missing values
df.isnull().sum()

Incident         0
Casualties       0
Total Dead       0
Crew             0
Passengers       0
Ground           0
Notes            0
Type             0
Reason           0
Location         0
Country          0
Phase            0
Date             0
Latitude         0
Longitude        0
Circumstances    0
dtype: int64

In [220]:
df.iloc[37]['Crew']

''

In [221]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 58 entries, 0 to 57
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype
---  ------         --------------  -----
 0   Incident       58 non-null     str  
 1   Casualties     58 non-null     str  
 2   Total Dead     58 non-null     str  
 3   Crew           58 non-null     str  
 4   Passengers     58 non-null     str  
 5   Ground         58 non-null     str  
 6   Notes          58 non-null     str  
 7   Type           58 non-null     str  
 8   Reason         58 non-null     str  
 9   Location       58 non-null     str  
 10  Country        58 non-null     str  
 11  Phase          58 non-null     str  
 12  Date           58 non-null     str  
 13  Latitude       58 non-null     str  
 14  Longitude      58 non-null     str  
 15  Circumstances  58 non-null     str  
dtypes: str(16)
memory usage: 7.4 KB


In [222]:
def clean_and_infer(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()

    # 去掉字符串列首尾空白
    for col in out.select_dtypes(include="object").columns:
        out[col] = out[col].str.strip()

    # 转日期
    if "Date" in out.columns:
        out["Date"] = pd.to_datetime(out["Date"], errors="coerce")

    # 转整数（可空 Int64）
    for col in ["Total Dead", "Crew", "Passengers", "Ground"]:
        if col in out.columns:
            out[col] = pd.to_numeric(out[col], errors="coerce").astype("Int64")

    # 转浮点（Float64）
    for col in ["Latitude", "Longitude"]:
        if col in out.columns:
            out[col] = pd.to_numeric(out[col], errors="coerce").astype("Float64")

    return out

df = clean_and_infer(df)

df.dtypes

/var/folders/6s/bb6mgp0x50q3fjlqfsfpffjm0000gn/T/ipykernel_68489/898140417.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  for col in out.select_dtypes(include="object").columns:


Incident                    str
Casualties                  str
Total Dead                Int64
Crew                      Int64
Passengers                Int64
Ground                    Int64
Notes                       str
Type                        str
Reason                      str
Location                    str
Country                     str
Phase                       str
Date             datetime64[us]
Latitude                Float64
Longitude               Float64
Circumstances               str
dtype: object

In [223]:
df.isnull().sum()

Incident         0
Casualties       0
Total Dead       0
Crew             1
Passengers       0
Ground           0
Notes            0
Type             0
Reason           0
Location         0
Country          0
Phase            0
Date             1
Latitude         0
Longitude        0
Circumstances    0
dtype: int64

pandas.DataFrame.isnull()（或者 df.isna()）默认只会把 真正的缺失值（NaN, None, NaT 等）当作缺失，不会把空字符串 "" 识别为缺失。


In [224]:

dft = pd.DataFrame({
    "a": [1, np.nan, ""]
})

print(dft.isnull())

       a
0  False
1   True
2  False


In [225]:
# handle missing values
# first see those rows
df[df.isnull().any(axis=1)]

,Incident,Casualties,Total Dead,Crew,Passengers,Ground,Notes,Type,Reason,Location,Country,Phase,Date,Latitude,Longitude,Circumstances
37,Incident Avianca Flight 011 involving a Boeing...,Extremely High,181,<NA>,162,0,Some survivors,COM,Accident,Madrid - Spain,Spain,APR,1983-11-27,40.416775,-3.70379,Bad Visibility by Night
47,Incident Union de Transportes Aériens Flight 7...,Extremely High,170,14,156,0,No survivors,INB,Attack,Tenere - Niger,Niger,ENR,NaT,18.954614,10.913461,Bad Visibility by Night


In [226]:
df[df.isnull().any(axis=1)].loc[df["Incident"].str.contains("Boeing", na=False), "Incident"].tolist()

['Incident Avianca Flight 011 involving a Boeing 747-283B in 1983']

In [227]:
# find incident include Boeing
df[df["Incident"].str.contains("Boeing 747", na=False)][["Incident", "Crew"]]

,Incident,Crew
2,Incident Pan Am Flight 1736 and KLM Flight 480...,23
4,Incident Saudi Arabian Flight 763 and involvin...,33
6,Incident Air India Flight 182 involving a Boei...,22
11,Incident Pan Am Flight 103 involving a Boeing ...,16
12,Incident Korean Air Lines Flight 007 involving...,23
20,Incident TWA Flight 800 involving a Boeing 747...,18
23,Incident Korean Air Flight 801 involving a Boe...,22
24,Incident China Airlines Flight 611 involving a...,19
27,Incident Air India Flight 855 involving a Boei...,23
37,Incident Avianca Flight 011 involving a Boeing...,<NA>


In [228]:
# simply remove the rows with missing values
df = df.dropna()
df[df.isnull().any(axis=1)]

,Incident,Casualties,Total Dead,Crew,Passengers,Ground,Notes,Type,Reason,Location,Country,Phase,Date,Latitude,Longitude,Circumstances


In [229]:
# 3. Check if there are any columns can be merged together? Apply the merging operation.
if "Crew" in df.columns and "Passengers" in df.columns:
    df["Total"] = df["Crew"] + df["Passengers"]
    # 4. Remove the column(s) that contains duplicated information after merging.
    df.drop(columns=["Crew", "Passengers"], inplace=True)

df.head()

,Incident,Casualties,Total Dead,Ground,Notes,Type,Reason,Location,Country,Phase,Date,Latitude,Longitude,Circumstances,Total
0,Incident American Airlines Flight 11 involving...,Extremely High,1692,1600,No survivors,INH,Attack,New York - New York - US,US,ENR,2001-09-11,40.714353,-74.005973,Good Visibility by Day,92
1,Incident United Airlines Flight 175 involving ...,Extremely High,965,900,No survivors,INH,Attack,New York - New York - US,USA,ENR,2001-09-11,40.714353,-74.005973,Good Visibility by Day,65
2,Incident Pan Am Flight 1736 and KLM Flight 480...,Extremely High,583,0,Some survivors,COM,Accident,Tenerife - Spain,Spain,TOF,1977-03-27,28.291564,-16.62913,Bad Visibility by Day,583
3,Incident Japan Airlines Flight 123 involving a...,Extremely High,520,0,Some survivors,COM,Accident,Ueno - Japan,Japan,ENR,1985-08-12,35.708946,139.774268,Bad Visibility by Night,520
4,Incident Saudi Arabian Flight 763 and involvin...,Extremely High,349,0,No survivors,,Accident,Charkhi Dadri - India,India,ENR,1996-11-12,28.6,76.2667,Bad Visibility by Night,349


In [230]:
# 5. Find out a subset that records air crashes with survivors.
df_survivors = df[df["Total"] > df["Total Dead"]]
df_survivors

,Incident,Casualties,Total Dead,Ground,Notes,Type,Reason,Location,Country,Phase,Date,Latitude,Longitude,Circumstances,Total
29,Incident Aeroflot Flight 7425 involving a Tupo...,Extremely High,0,0,No survivors,COM,Accident,Uchkuduk - Uzbekistan,Uzbekistan,ENR,1985-07-10,42.156667,63.555556,Bad Visibility by Night,200
56,Incident China Northwest Airlines Flight 2303 ...,Extremely High,150,0,No survivors,COM,Accident,Xian - China,China,ENR,1984-06-06,34.341568,108.940175,Bad Visibility by Night,160
